lib/access-context.js

```diff
@@ -126,9 +130,29 @@ AccessContext.prototype.addPrincipal = function(principalType, principalId, prin
 AccessContext.prototype.getUserId = function() {
   for (var i = 0; i < this.principals.length; i++) {
     var p = this.principals[i];
+    if (p.type === Principal.APP || p.type === Principal.ROLE || p.type == Principal.SCOPE) {
+      continue;
+    }
+    // the principalType must either be 'USER'
     if (p.type === Principal.USER) {
       return p.id;
     }
+    // or permit to resolve a valid user model
+    var userModel = this.registry.findModel(p.type);
+    if (!userModel) continue;
+    if (userModel.modelName === 'User') {
+      return p.id;
+    }
+    var modelBase = userModel.definition.settings.base;
+    while (modelBase) {
+      // make sure modelBase is a model instance
+      modelBase = typeof(modelBase === 'string') ? this.registry.findModel(modelBase) : modelBase;
+      if (modelBase.modelName === 'User') {
+        return p.id;
+      }
+      // iterate on the upper base model
+      modelBase = modelBase.definition.settings.base;
+    }
```

Comment:
> @batjos: Uff, I am actually confused, what is the purpose of the code here and below, where you are crawling model hierarchy?

you commented on an earlier implementation:
> @bajtos : This does not work with a deeper hierarchy (Customer extends BaseUser extends built-in User).

to what i replied : 
> @ebarault : you're right, i noted that too. we should test recursively up to the top User model to check if we ultimately get the built-in User model.

I changed the implementation accordingly and added a [test](https://github.com/strongloop/loopback/pull/2971/files#diff-e47f5e866ee5dd9fcbd37d61213a7210R233)